api/python/tests/integration/test_packages.py

```diff
@@ -666,6 +666,61 @@ def test_s3_set_dir(self):
 
             list_object_versions_mock.assert_called_with('bucket', 'foo/')
 
+    def test_set_dir_update_policy(self):
+        """Verify building a package with update policy. """
+
+        # Test for local src dir
+        foodir = pathlib.Path("foo_dir")
+        bazdir = pathlib.Path("baz_dir")
+        foodir.mkdir()
+        bazdir.mkdir()
+        with open(foodir / 'bar', 'w') as fd:
+            fd.write(fd.name)
+        with open(bazdir / 'bar', 'w') as fd:
+            fd.write(''.join([str(i) for i in range(100)]))
+
+        pkg = Package()
+        pkg = pkg.set_dir("/", ".", meta={'name': 'test_meta'})
+        assert 'foo_dir' in pkg.keys()
+        existing_bar_size = pkg['foo_dir/bar'].size
+
+        # existing update policy
+        pkg = pkg.set_dir("foo_dir", bazdir, update_policy='existing')
+        assert 'foo_dir' in pkg.keys()
+        assert existing_bar_size == pkg['foo_dir/bar'].size
+
+        # incoming update policy
+        pkg = pkg.set_dir("foo_dir", bazdir)
+        assert 'foo_dir' in pkg.keys()
+        assert existing_bar_size != pkg['foo_dir/bar'].size
+
+        # verify non existing update policy raises value error
+        with pytest.raises(ValueError):
+            pkg.set_dir("foo_dir", foodir, update_policy='invalid_policy')
+
+        # Test for s3 src dir
+        with patch('quilt3.packages.list_object_versions') as list_object_versions_mock:
+            pkg = Package()
+
+            list_object_versions_mock.return_value = ([
+                dict(Key='foo/a.txt', VersionId='xyz', IsLatest=True, Size=10),
+            ], [])
+
+            pkg.set_dir('', 's3://bucket/foo/', meta={'name': 'test_meta'})
+            assert pkg['a.txt'].get() == 's3://bucket/foo/a.txt?versionId=xyz'
+
+            list_object_versions_mock.return_value = ([
```

Comment:
It's cleaner to do this in a separate test or in a loop. You can do `with patch` and loop over the various update policies.

Reply:
this is fine but you need to check that these mocks fired; see `.assert_has_calls` and check `.call_count == N`